core/src/main/java/org/apache/calcite/sql2rel/TopDownGeneralDecorrelator.java

```diff
@@ -0,0 +1,1009 @@
+                leftInfo,
+                rightInfo,
+                correlate,
+                unnestedRel,
+                correlate.getJoinType());
+        List<RexNode> projects
+            = builder.fields(new ArrayList<>(unnestQuery.oldToNewOutputs.values()));
+        unnestedRel = builder.project(projects).build();
+      }
+      return unnestedRel;
+    } else {
+      for (int i = 0; i < rel.getInputs().size(); i++) {
+        rel.replaceInput(i, correlateElimination(rel.getInput(i)));
+      }
+    }
+    return rel;
+  }
+
+  /**
+   * Generate the domain of the free variables D.
+   *
+   * @param newLeft   the left side (without correlation) of Correlate
+   * @param leftInfo  the UnnestQuery of the left side of Correlate
+   * @return  the domain of the free variables D
+   */
+  private DedupFreeVarsNode generateDedupFreeVarsNode(RelNode newLeft, UnnestQuery leftInfo) {
+    List<Integer> columnIndexes = new ArrayList<>();
+    for (CorDef corDef : corDefs) {
+      int fieldIndex = requireNonNull(leftInfo.corDefOutputs.get(corDef));
+      columnIndexes.add(fieldIndex);
+    }
+    List<RexNode> inputRefs = builder.push(newLeft)
+        .fields(columnIndexes);
+    RelNode rel = builder.project(inputRefs).distinct().build();
+    return new DedupFreeVarsNode(rel);
+  }
+
+  /**
+   * Detects whether existing correlated expressions in the RelNode (according to corDefs).
+   *
+   * @param rel RelNode
+   * @return true when there are correlated expressions
+   */
+  private boolean detectCorrelatedExpressions(RelNode rel) {
+    boolean hasCorrelation = false;
+    for (RelNode input : rel.getInputs()) {
+      hasCorrelation |= detectCorrelatedExpressions(input);
+    }
+    if (!hasCorrelation) {
+      RelOptUtil.VariableUsedVisitor variableUsedVisitor =
+          new RelOptUtil.VariableUsedVisitor(null);
+      rel.accept(variableUsedVisitor);
+      Set<CorrelationId> corrIdSet
+          = corDefs.stream()
+          .map(corDef -> corDef.corr)
+          .collect(ImmutableSet.toImmutableSet());
+      hasCorrelation =
+          !variableUsedVisitor.variables.isEmpty()
+              && !Collections.disjoint(corrIdSet, variableUsedVisitor.variables);
+    }
+    hasCorrelatedExpressions.put(rel, hasCorrelation);
+    return hasCorrelation;
+  }
+
+  /**
+   * Create the new join condition after decorrelating.
+   *
+   * @param oriCondition        original Correlate/Join condition
+   * @param leftInfo            UnnestQuery of the left side
+   * @param rightInfo           UnnestQuery of the right side
+   * @param needNatureJoinCond  whether need to add the natural join condition for domain D
+   * @return the new join condition
+   */
+  private RexNode createUnnestedJoinCondition(
+      RexNode oriCondition,
+      UnnestQuery leftInfo,
+      UnnestQuery rightInfo,
+      boolean needNatureJoinCond) {
+    // create a virtual inner join and its UnnestQuery to help rewrite the
```

Comment:
I am guessing that "Virtual" means "temporary", which is not used in the final plan.
You are doing this only to reuse the existing visitor CorrelatedExprRewriter.